app/dial.go

```diff
@@ -39,3 +39,21 @@ func makeNodeDialFunc(config *tls.Config) client.DialFunc {
 
 	return dial
 }
+
+func extDialFuncWithProxy(dialFunc client.DialFunc) client.DialFunc {
+	return func(ctx context.Context, addr string) (net.Conn, error) {
```

Comment:
Why does the function here accept a ctx, but doesn't pass it to the `go proxy()` function call inside it?

Reply:
ah thanks I missed that.

Reply:
Hmm, actually it appears the parent context is already done by the time the dialfunc is run, which seems odd. Looking to find out why.

Reply:
Ok so the parent context starts from the `connectWithDial` method in `internal/bindings/server.go`, which is called from `c`, and this is what executes the DialFunc. The context isn't cancelled until the DialFunc returns, but since we're running the proxy in a goroutine that we want to run as long as there's data between the connections,  we don't want the context to be cancelled by the caller.

Reply:
OK makes sense, so one is a connect timeout CTX and one is a cancelation context. 

What concerns me is why in one call to proxy() a cancelation context is passed as expected, but here the background context is passed meaning it won't get cancelled.

Reply:
Ok, I added a context to the `Node` struct that starts/stops the dqlite server, which is then assigned to a variable in the `c` bindings when the server starts, and is cancelled when the server stops. 

This context becomes the parent context for the dial function, which in turn is only cancelled on a timeout or if the dial function returns any errors.